clients/drcachesim/analysis_tool.h

```diff
@@ -227,16 +223,34 @@ template <typename RecordType> class analysis_tool_tmpl_t {
         return nullptr;
     }
     /**
-     * Invoked by the framework to combine the two \p interval_state_snapshot_t
-     * objects pointed by the given inputs. This is useful in the parallel mode
-     * of analyzer, where each trace shard is processed in parallel. In this mode,
-     * the \p interval_state_snapshot_t from different shards that map to the same
-     * final whole-trace interval need to be combined into one. This API must not
-     * return a nullptr.
+     * Invoked by the framework to combine the shard-local \p interval_state_snapshot_t
+     * objects pointed at by \p last_shard_snapshots, to create the combined
+     * \p interval_state_snapshot_t for a whole-trace interval. This is useful in the
+     * parallel mode of the analyzer, where each trace shard is processed in parallel. In
+     * this mode, the framework creates the \p interval_state_snapshot_t for each
+     * whole-trace interval, one by one. For each interval, it invokes
+     * combine_interval_snapshots() with the as-then last interval snapshots from each
+     * shard (in \p last_shard_snapshots), and also information about whether each of
+     * those snapshots was observed in the current interval (in \p
+     * observed_in_cur_interval). \p last_shard_snapshots will contain at least one
+     * non-null element. Tools must not return a nullptr. Tools must not modify the
+     * provided snapshots.
+     *
+     * \p observed_in_cur_interval helps the tool to determine which of the provided
+     * \p last_shard_snapshots are relevant to it.
```

Comment:
> Or, aren't the interval ids globally consistent 

They aren't. Hard to have consistent `interval_id` in the parallel mode since it depends on the first timestamp of each shard.

> could we add this bool inside interval_state_snapshot_t.

interval_state_snapshot_t already has an `interval_end_timestamp` which the tool can use to figure out which shard snapshots were observed in the most recent interval (the ones with the largest value of `interval_end_timestamp`). I added it as an explicit arg as that seemed simpler for tools. Do you recommend we remove the extra bool vector and let the tools figure out as above?

Reply:
Oh there's a simpler way. We can just pass the interval_end_timestamp to the combine function, and then the tool can figure out the recent ones based on the timestamp in the snapshot.

Reply:
Done.